### ISSUE TITLE
Suggestion Layer Scale

### ISSUE THREAD

**Issue by R0CH1TV** [kamilburda/batcher#16]
Hello, maybe you can add option to resize scale based on canvas size...

Example

My canvas size is 40x20
![image](https://github.com/user-attachments/assets/ae6fec45-ba34-423b-8abf-351672c422b8)

My layers have different size
Original size
Layer1: 548x141
![image](https://github.com/user-attachments/assets/b0088d2c-9a2b-4d58-845f-76c4d647297d)
Layer2: 236x182
![image](https://github.com/user-attachments/assets/6454cdf4-2853-4497-8b94-857aef8cfc9b)

Resized
Layer1: 40x10 (fit on canvas size)
![image](https://github.com/user-attachments/assets/1aac64f9-eecc-497c-89aa-dd9fb83f7fe8)

Layer2: 40x31 (exceed canvas size on height 31px) so you need resize by Height
![image](https://github.com/user-attachments/assets/b2447c36-c5e6-4640-a739-1cd480eaa260)

Resize by height then (keeping scale ratio)
![image](https://github.com/user-attachments/assets/f032b768-883c-4737-bd92-285cd99a5cca)

So basically you always try to resize by the largest value, but if the height or width still exceeds the canvas size, you resize whatever is exceeding it, but always maintaining the proportion.

**Comment by kamilburda** [comment]
As I understand it, you're requesting "scale to fit" while also preserving the aspect ratio. I'll add this option to the next version. 🙂

**Comment by R0CH1TV** [comment]
> As I understand it, you're requesting "scale to fit" while also preserving the aspect ratio. I'll add this option in the next version. 🙂

Yes, basically... but you always try to resize it by the largest value first, but if it is still larger than the canvas sizes, you try the other value.

Canvas Size: 40x20

Example:

Width: 500
Height: 350

but if you put width at : 40 and the height is 25
you need to resize the height to 20 and the width will be 35

that was just an example, sorry for the somewhat confusing explanation.

**Comment by kamilburda** [comment]
> Sorry for too many questions, and this annoying question

No worries 🙂

> you have any ETA for new version?

I am in the middle of adding a [major new feature to the plug-in](https://github.com/kamilburda/batcher/issues/10), and that will take quite a while - possibly 1-2 months.

However, scaling to fit and keeping aspect ratio are actually quite useful features, and can be done fairly quickly, and I suppose I could drop a minor release before 1.0. I can't promise any exact ETA due to other commitments, though (such as my job 😅), but I'd wager it'd be some time before the holidays.

**Comment by R0CH1TV** [comment]
> > Sorry for too many questions, and this annoying question
> 
> No worries 🙂
> 
> > you have any ETA for new version?
> 
> I am in the middle of adding a [major new feature to the plug-in](https://github.com/kamilburda/batcher/issues/10), and that will take quite a while - possibly 1-2 months.
> 
> However, scaling to fit and keeping aspect ratio are actually quite useful features, and can be done fairly quickly, and I suppose I could drop a minor release before 1.0. I can't promise any exact ETA due to other commitments, though (such as my job 😅), but I'd wager it'd be some time before the holidays.

I understand you !
would be very nice if you able to drop this feature when you have time this will help me a lot xD
but thank you !

**Comment by kamilburda** [comment]
I just released a new version containing the new scaling features. To fit each layer to the image canvas:
1. Set the New width to 100 and the Unit for width to `% of image width`.
2. Set the New height to 100 and the Unit for height to `% of image height`.
3. Check `Scale to fit`.

Of course, if you need to specify other dimensions representing the canvas, such as 40 x 20, that can be done as well. Simply change the units to pixels and set the New width and height to 40 and 20, respectively.

However, there may be one problem with this approach - after the scaling, some parts of the layer may be displayed outside the image canvas. If you only want to export the layers using their dimensions (and not the image dimensions), simply check `Use layer size` if not already. Otherwise, you'll need to set layer offsets via a custom procedure (`gimp-layer-set-offsets`). However, if you need to use different values than pixels (e.g. center the layer with respect to the image canvas), this cannot currently be done, unless you find a custom script/plug-in that can then be inserted as a custom procedure in Batcher.

**Comment by R0CH1TV** [comment]
> I just released a new version containing the new scaling features. To fit each layer to the image canvas:
> 
> 1. Set the New width to 100 and the Unit for width to `% of image width`.
> 2. Set the New height to 100 and the Unit for height to `% of image height`.
> 3. Check `Scale to fit`.
> 
> Of course, if you need to specify other dimensions representing the canvas, such as 40 x 20, that can be done as well. Simply change the units to pixels and set the New width and height to 40 and 20, respectively.
> 
> However, there may be one problem with this approach - after the scaling, some parts of the layer may be displayed outside the image canvas. If you only want to export the layers using their dimensions (and not the image dimensions), simply check `Use layer size` if not already. Otherwise, you'll need to set layer offsets via a custom procedure (`gimp-layer-set-offsets`). However, if you need to use different values than pixels (e.g. center the layer with respect to the image canvas), this cannot currently be done, unless you find a custom script/plug-in that can then be inserted as a custom procedure in Batcher.

Really good job working amazing...
But like you said now my problem layers not centralized on canvas anymore,
it's hard to add option do "centralized" on canvas after resize?

**Comment by kamilburda** [comment]
I have a Python script available for aligning layers: https://github.com/kamilburda/gimp-align-layer

I just updated it to work with GIMP 3. You can install it as any other GIMP plug-in/script. First restart GIMP so that the plug-in is registered, then in Batcher, find and add the `python-fu-align-layers` as a custom procedure. There you can adjust how to align the current layer - keep, left/top, center, right/bottom. You can apply additional offsets if needed.

**Comment by R0CH1TV** [comment]
> I have a Python script available for aligning layers: https://github.com/kamilburda/gimp-align-layers
> 
> I just updated it to work with GIMP 3. You can install it as any other GIMP plug-in/script. First restart GIMP so that the plug-in is registered, then in Batcher, find and add the `python-fu-align-layers` as a custom procedure. There you can adjust how to align the current layer - keep, left/top, center, right/bottom. You can apply additional offsets if needed.
> 
> You can ignore the reference layer (unless you inserted e.g. a background layer and would want to align the current layer against the background).

Sure i added and selected but i getting error

![image](https://github.com/user-attachments/assets/57755d9b-cc68-4d8a-ae97-88c387a3a87b)

My settings

![image](https://github.com/user-attachments/assets/040e077a-b2a3-4fd0-805d-53996d8a6a9a)

**Comment by kamilburda** [comment]
That's the `plug-in-align-layers` plug-in. You'll need to add `python-fu-align-layers`. I should probably have named my script differently 😅

**Comment by R0CH1TV** [comment]
> That's the `plug-in-align-layers` plug-in. You'll need to add `python-fu-align-layers`. I should probably have named my script differently 😅

![image](https://github.com/user-attachments/assets/9f0e4a51-bf79-4b20-a4af-25b942feb5cc)


LOL problem i can't find that for some reason....
though console doesn't show any errors to load the script ,

![image](https://github.com/user-attachments/assets/5511ceea-3c99-4dcd-9912-56639ab71291)


Folder
![image](https://github.com/user-attachments/assets/e5f41c07-25ec-44c0-b04b-48bf88d5c90f)

**Comment by kamilburda** [comment]
Check if you have the `scripts` folder among the plug-ins folders. Also, you need to restart GIMP if you did not do so already.

**Comment by R0CH1TV** [comment]
worked thank u !!

**Comment by kamilburda** [comment]
Great, I'm considering this resolved then. 🙂
